crates/core/src/host/host_controller.rs

```diff
@@ -101,249 +194,649 @@ impl From<&EventStatus> for ReducerOutcome {
+    ) -> anyhow::Result<watch::Receiver<ModuleHost>> {
+        let program_hash = database.program_bytes_address;
+
+        // Try a read lock first.
+        {
+            let guard = self.acquire_read_lock(instance_id).await;
+            if let Some(host) = &*guard {
+                trace!("cached host {}/{}", database.address, instance_id);
+                // The stored program must be equal to the one in `Database`.
+                guard_program_hash(host.db(), Some(program_hash))?;
+                return Ok(host.module.subscribe());
+            }
+        }
+
+        // We didn't find a running module, so take a write lock.
+        // Since [`tokio::sync::RwLock`] doesn't support upgrading of read locks,
+        // we'll need to check again if a module was added meanwhile.
+        let mut guard = self.acquire_write_lock(instance_id).await;
+        if let Some(host) = &*guard {
+            trace!("cached host {}/{} (lock upgrade)", database.address, instance_id);
+            // The stored program must be equal to the one in `Database`.
+            guard_program_hash(host.db(), Some(program_hash))?;
+            return Ok(host.module.subscribe());
+        }
+
+        trace!("launch host {}/{}", database.address, instance_id);
+        let host = self.try_init_host(database, instance_id).await?;
+        // The stored program must be equal to the one in `Database`.
+        guard_program_hash(host.db(), Some(program_hash))?;
+
+        let rx = host.module.subscribe();
+        *guard = Some(host);
+
+        Ok(rx)
     }
 
-    /// Exit and delete the module corresponding to the provided instance id.
+    /// Run a computation on the [`RelationalDB`] of a [`ModuleHost`] managed by
+    /// this controller, launching the host if necessary.
     ///
-    /// Note that this currently does not delete any physical data.
-    pub async fn delete_module_host(
-        &self,
-        _fence: u128,
-        worker_database_instance_id: u64,
-    ) -> Result<(), anyhow::Error> {
-        // TODO(kim): If the delete semantics are to wipe all state from
-        // persistent storage, `_fence` is not needed. Otherwise, we will need
-        // to check it against the stored value to be able to order deletes wrt
-        // other lifecycle operations.
-        //
-        // Note that currently we don't delete the persistent state, but also
-        // imply that a deleted database cannot be resurrected.
-        if let Some(host) = self.take_module_host(worker_database_instance_id) {
-            host.exit().await;
-        }
-        Ok(())
+    /// An error is returned if the host's program does not match the hash given
+    /// in [`Database`].
+    ///
+    /// If the computation `F` panics, the host is removed from this controller,
+    /// releasing its resources.
+    #[tracing::instrument(skip_all)]
+    pub async fn using_database<F, T>(&self, database: Database, instance_id: u64, f: F) -> anyhow::Result<T>
```

Comment:
Less than pleasant, arguably.
Can't think of a better way, though, because relational db is not `UnwindSafe`.

Reply:
You could just use `AssertUnwindSafe` - we don't much care about unwinding anyway

Reply:
Maybe, but that would still mean to catch unwind in every single method of relational db. That seems silly.

Reply:
Just out of pure curiosity, what makes `RelationalDB` not `UnwindSafe`? Keep in mind I know very little about Rust stack unwinding

Reply:
`parking_lot` locks used in the datastore keep their data in a literal `UnsafeCell`. Catching a panic means code can continue to run (e.g. destructors), and the compiler cannot prove anything about the memory safety of an `UnsafeCell` (because that is what it's for :)). At least that's how I understand it.

I believe what Noa is saying is that our goal is to _drop_ the whole `Host` when there is a panic, so unless we have `Drop` impls which do silly things, this should be safe.

Thinking about it some more, it might be possible and sufficient to only catch panicking of `durability.append_tx` in `RelationalDB::commit_tx`. This is the place where we _deliberately_ panic, and once the method panicked it will continue to panic (to signal that the durability layer is in an irrecoverably broken state). Any other panic (e.g. bad unwrap) will make the current request panic, but the next one is likely to succeed.

This still requires to assert unwind safety, so I'll try it in a follow up for easier review.